proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+``ImplicitStagePersistence`` (provided the latter appears later than the former,
+so it overrides the implied ``NoImplicitStagePersistence``). This allows
+``splice`` and ``quote`` imports to be used, but ``ImplicitStagePersistence``
+still allows cross-stage persistence (and thus the compiler must still be
+pessimistically assume all modules are needed at all stages). This combination
+is supported to allow gradual migration of code bases following the change, and
+for corner cases such as code generation, where the programmer may wish to use
+the syntax of ``splice`` and ``quote`` imports without obliging the whole module
+to be level-correct.
+
+
+Syntax for imports
+##################
+
+Under ``ExplicitLevelImports``, the syntax for imports becomes::
+
+  importdecl :: { LImportDecl GhcPs }
+     : 'import' maybe_src maybe_safe optsplice optqualified maybe_pkg modid optqualified maybeas maybeimpspec
+
+  optsplice :: { LImportStage }
+     : 'splice' { SpliceStage }
+     | 'quote'  { QuoteStage  }
+     |          { NormalStage }
+
+
+The ``splice`` or ``quote`` keyword appears before the ``qualified`` keyword but after ``SOURCE``
+and ``SAFE`` pragmas.
+
+
+Name resolution and ambiguous variables
+#######################################
+
+Name resolution ("renaming") does not take account of the level at which an
+identifier was imported when disambiguating ambiguous names, even though this is
+sometimes more conservative than necessary.  For example, the following program
+is rejected::
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+
+  import A ( x )
+  import splice B ( x )
+
+  foo = $( x ) x
+
+In this case, there is in principle no ambiguity because ``A.x`` isn't allowed
+to be used in the top-level splice, and ``B.x`` isn't allowed to be used outside
+the splice.  Thus the only disambiguation that will pass the type-checker is::
+
+  foo = $( B.x ) A.x
+
+We choose to reject this disambiguation to keep the design simple and prevent
+any confusion about what is in scope. This position is conservative, and can be
+relaxed in the future if more flexibility appears worthwhile.
+
+A positive consequence of the current design is that if a program is accepted
+with ``ExplicitLevelImports``, it will be accepted after erasing all
+``splice``/``quote`` keywords and using ``ImplicitStagePersistence`` instead of
+``ExplicitLevelImports``.
+
+
+Exports
+#######
+
+Under ``NoImplicitStagePersistence``, modules may export bindings only if they
+are available at level 0.
+
+For example, the following is rejected::
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+
+  module M (oops) where  -- Error: oops imported at level -1 but used at level 0
+    import splice N ( oops )
+
+
+Class instance resolution
+#########################
+
+Class instances carry a level, much like identifiers, and must be used at the
+correct level.  This will be enforced by the type-checker under
+``NoImplicitStagePersistence``:
```

Comment:
In the proposal, top-level bindings are bound at level 0 and follow all the same rules as level 0 imports. 

Currently for instances defined in the same module the compiler implements the same checks as for variables.

You are not allowed to use an instance (level 0) from the same module in a top-level splice (level -1). 

For untyped brackets, the body is not typechecked so you never end up requesting to use instances at level 1.

Reply:
Additional checks for instances are not straightforward implementation wise because instances are not explicitly exported/imported from modules, they are brought into scope implicitly. In order to resolve this we compute the "visible transitive closure", which then tells us which module could provide instances which are visible and then check the instance is from one of those.